Feature: LR-TDDFT

maki49 opened the pull request:
The main part of #2460.
### Feature
Construct (Hamilt(Hcontainer)+Operator(Gint+LibRI))  and solve (lapack/dav/dav_subspace) Casida equation $AX=\Omega X$ to get excitation energy $\Omega$ and amplitudes $X$, then use them to calculate absorption spectra.
The main workflow is implemented in `ESolver_LR`.
- Supported xc kernel: rpa(no xc), lda, pbe, hf, hse
- Both after-KS (by move constructors) and from-scratch (by reading KS eigenstates)

### Remaining Problems
- large negative eigenvalues with `dav_subspace` and `psi_norm < 0` with `dav`, which prohibits too-many k-points
- nspin=2 is not supported (spin-flip transition is not taken into account), and the formulas of triplet in nspin=1 framework is still need to be discussed

maki49 committed:
add operator=(T&&) in Grid_Technique and its 4 base classes

remove getters and n_unocc, use shared_ptr

maki49 committed:
X-index: diagonal arrangement and tests

small fix after merge

Para2D distribute X

maki49 committed:
dm_trans_serial

build directory structure

cal_dm_trans: pzgemm and tests

maki49 committed:
Framework: kernel, pot, operator, hamilt

AX-cVc (including UT), fix PW-compile

complex dm-trans

complex AX

unify dm_trans algotithm

container::Tensor

Refactor: minor interface in Gint

Procedure: build gamma_only AX

small fixes

band-first AX

band-first dm_trans

update dm_trans and AX UT to nstate=2

param xc_kernel & fix gint-call and Kernels

add hsolver (tmp)

maki49 committed:
multi-k AX routine based on HContainer & DensityMatrix

fix a nullptr in Psi(T*)

fix: paraX, paraMat, HContainer&DM, Davidson-precondition

a common pointer constructor of Psi

psi wrapperL: k1<->bfirst

small fixes

fix act interface: nbands<=psi_in.get_nbands()

get_act_type() for each operator

refactor: exx dm

fix out_dm

maki49 committed:
OperatorLREXX

generalize gint_kernel_rho

maki49 committed:
OperatorLRDiag

real-valued CG

fix hartree & eigensolver paras for lr

dm symmetrization

maki49 committed:
from-scratch LR

matrix-lapack, reuse DM, fix wf-sign-dependence

maki49 committed:
spectrum: absorption & transition analysis

add an example

small fixes

maki49 committed:
lapack method: enable multi-k parallel

modify the conjugateness of c in DM_trans

fix negative eigenvalue of multi-k

maki49 committed:
GGA-kernel & refactor & example-Si

unify nks

respective gint for real and imag part in OperatorHxc

refactor print and modify init

scal with 1/Nk and transpose dm_trans

imag part in absorption spectra

hse kernel & IO eigenstate & spectrum-only

fix after rebase

remove a non-used parameter in

fix after rebase 2

fix dipole magnitude

fix k-first in dav_subspace (still bug)

fix V in AX

xc_kernel: replace matrix with vector

refactor XC-kernel and fix the missing e2 (GGA still bug)

fix invalid lr_ethr

fix GGA kernel

make the broadening eta an input parameter

fix after rebase

fix after rebase (dav_subspace still negative eig)

fix exx compile and Makefile

default move Parallel_2D

move LR-init from driver to esolver; rm some move constructors

mohanchen commented on source/driver_run.cpp:
```diff
@@ -2,86 +2,134 @@
-    //! 3: initialize Esolver and fill json-structure 
-    p_esolver->before_all_runners(INPUT, GlobalC::ucell);
+    //! 3: initialize Esolver and fill json-structure
+#ifdef __LCAO
```
It seems that this part of code belong to somewhere else.

maki49 replied:
I've moved this part into `esolver.cpp` int the newest commit.

mohanchen commented on source/driver_run.cpp:
```diff
@@ -2,86 +2,134 @@
     p_esolver->after_all_runners();
-    ModuleESolver::clean_esolver(p_esolver);
 
+#ifdef __LCAO
```
It seems that this part of code belong to somewhere else.

jinzx10 commented on source/module_basis/module_ao/parallel_2d.h:
```diff
@@ -16,6 +16,8 @@ class Parallel_2D
     Parallel_2D() = default;
     ~Parallel_2D() = default;
 
+    Parallel_2D& operator=(Parallel_2D&& rhs);
```
Since the members of this class are either POD types or std::vector, an easier way to implement the move assignment operator is
```suggestion
    Parallel_2D& operator=(Parallel_2D&& rhs) = default;
```

maki49 replied:
Changed. It works. Thanks for your refactor.

Qianruipku commented on source/driver_run.cpp:
```diff
@@ -2,86 +2,134 @@
+    if (GlobalV::ESOLVER_TYPE == "lr")
+        // use constructor rather than Init function to initialize reference (instead of pointers) to ucell
+        if (GlobalV::GAMMA_ONLY_LOCAL)
+            p_esolver = new ModuleESolver::ESolver_LRTD<double, double>(INPUT, GlobalC::ucell);
```
Please move this part to init_esolver function. This cpp file should not include anything about LCAO.  This file should not have "#ifndef __LCAO" or "ifdef __LCAO".

Qianruipku replied:
driver should not depend on LCAO, it not depend on esolver

maki49 replied:
I don't think #ifdef-#endif means dependence, because the dependent header `esolver_lrtd_lcao` is also in #ifdef-#endif.
Also, there's another `__LCAO` in this file:
https://github.com/deepmodeling/abacus-develop/blob/develop/source/driver_run.cpp#L34-L39

maki49 replied:
My `ESolver_LRTD`'s constructor depends on an **initialized** `ucell`. 
I will move `init_esolver` after `ucell.setup_cell` and make it depend on `ucell`, which I think is a reasonable order and dependence. If you agree, then this part can be moved into `init_esolver`.

mohanchen replied:
Of course, every esolver depends on ucell

Qianruipku commented on source/driver_run.cpp:
```diff
@@ -2,86 +2,134 @@
+        // ModuleESolver::ESolver_KS_LCAO* p_esolver_lcao_tmp = dynamic_cast<ModuleESolver::ESolver_KS_LCAO<double,
+        // double>*>(p_esolver);
+        ModuleESolver::ESolver* p_esolver_lr = nullptr;
+        if (INPUT.gamma_only)
```
move this part to functions like init_esolver

maki49 replied:
This part (initializing the 2nd esolver) is called **after** `ucell.setup_cell` and all the functions except for `clean_esolver` of the 1st esolver, but `init_esolver` is called **before** them. 

If `init_esolver` can be moved after `ucell.setup_cell`, maybe it's possible to copy the 1st-esolver-functions into the `if("ks-lr")` branch of `init_esolver` (followed with the initialization of the 2nd esolver), and share the subsequent 2nd-esolver-functions here with other branches. -- What do you think of changing like this?

maki49 replied:
I've changed it according to the above statement in the newest commit.

Qianruipku commented on source/module_beyonddft/dm_trans/test/dm_trans_test.cpp:
```diff
@@ -0,0 +1,277 @@
+    std::vector<matsize> sizes{{2, 14, 9, 4}, {2, 20, 10, 7}};
+    int nstate = 2;
+    std::ofstream ofs_running;
+    int my_rank;
```
It is better to give my_rank, naos, nocc, nvirt initial values.

maki49 replied:
OK, done.

mohanchen commented on docs/advanced/input_files/input-main.md:
```diff
@@ -3743,3 +3752,63 @@ These variables are used to control the usage of PEXSI (Pole Expansion and Selec
+
+- **Type**: String
+- **Description**: The method to solve the Casida equation in LR-TDDFT.
+  - `dav`: Construct $AX$ and diagonalize the Hamiltonian matrix iteratively with Davidson algorithm.
```
Young need to briefly describe A and X.

mohanchen commented on docs/advanced/input_files/input-main.md:
```diff
@@ -3743,3 +3752,63 @@ These variables are used to control the usage of PEXSI (Pole Expansion and Selec
+- **Description**: The method to solve the Casida equation in LR-TDDFT.
+  - `dav`: Construct $AX$ and diagonalize the Hamiltonian matrix iteratively with Davidson algorithm.
+  - `lapack`: Construct the full $A$ matrix and directly diagonalize with LAPACK.
+  - `spectrum`: Caluclate absorption spectrum only without solving Casida equation. The `OUT.${suffix}/` directory should contain the
```
typo, Calculate, not 'Caluclate'

mohanchen commented on docs/advanced/input_files/input-main.md:
```diff
@@ -3743,3 +3752,63 @@ These variables are used to control the usage of PEXSI (Pole Expansion and Selec
+### lr_thr
+
+- **Type**: Real
+- **Description**: The convergence threshold of iterative diagonalization fo LR-TDDFT.
```
What's the unit? What's the convergence property, energy or density, etc.?

mohanchen commented on docs/advanced/input_files/input-main.md:
```diff
@@ -3743,3 +3752,63 @@ These variables are used to control the usage of PEXSI (Pole Expansion and Selec
+- **Description**: The number of virtual orbitals used in the LR-TDDFT calculation.
+- **Default**: 1
+
+### nstates
```
this name is too general, I suggest changing 'nstates' to 'lr_nstates' or something more specific

mohanchen commented on source/driver_run.cpp:
```diff
@@ -2,86 +2,134 @@
-#include "module_io/para_json.h"
 
+#ifdef __LCAO
+#include "module_beyonddft/esolver_lrtd_lcao.h"
```
this line should not appear in driver_run.cpp, everything should be contained in esolver

mohanchen commented on source/module_beyonddft/dm_trans/dm_trans_parallel.cpp:
```diff
@@ -0,0 +1,214 @@
+
+    std::vector<container::Tensor> dm_trans(
+        nks,
+        container::Tensor(DAT::DT_DOUBLE, DEV::CpuDevice, {pmat.get_col_size(), pmat.get_row_size()}));
```
what are DAT and DEV, where are they defined?

maki49 replied:
In `lr_util.h`: 
```
using DAT = container::DataType;
using DEV = container::DeviceType;
```

maki49 committed:
rebase, modify namespace and rename dir

Hamilt and HSolver are not members anymore

rename and refactor init_X

apply new input

mohanchen commented:
close to resolve some clang-format issues, please resubmit later.

mohanchen commented on source/module_hamilt_lcao/module_gint/gint_k.h:
```diff
@@ -13,7 +13,16 @@
+    /// @brief move operator for the next ESolver to directly use its infomation
+    /// @param rhs 
+    /// @return *this
+    Gint_k& operator=(Gint_k&& rhs);
```
I would suggest removing this line. We don't want operator unless necessary.

maki49 replied:
Move constructors are necessary in my new code. We've talked about this at the meeting on July 2nd (Tue.)

mohanchen commented on source/module_hamilt_lcao/module_gint/grid_bigcell.h:
```diff
@@ -15,10 +15,13 @@ class Grid_BigCell: public Grid_MeshCell
+        std::vector<std::vector<double>> tau_in_bigcell;
+
+        /// move operator for the next ESolver to directly use its infomation
+        Grid_BigCell& operator=(Grid_BigCell&& rhs) = default;
```
I would suggest removing this line. We don't want operator unless necessary.

mohanchen commented on source/module_hamilt_lcao/module_gint/grid_meshball.h:
```diff
@@ -9,9 +9,12 @@ class Grid_MeshBall : public Grid_BigCell
+        std::vector<std::vector<double>> meshball_positions;
+
+        /// move operator for the next ESolver to directly use its infomation
+        Grid_MeshBall& operator=(Grid_MeshBall&& rhs) = default;
```
I would suggest removing this line. We don't want operator unless necessary.

mohanchen commented on source/module_hamilt_lcao/module_gint/grid_meshcell.h:
```diff
@@ -27,7 +27,10 @@ class Grid_MeshCell: public Grid_MeshK
 
-	void set_grid_dim(
+    /// move operator for the next ESolver to directly use its infomation
+    Grid_MeshCell& operator=(Grid_MeshCell&& rhs) = default;
```
I would suggest removing this line. We don't want operator unless necessary.

mohanchen commented on source/module_hamilt_lcao/module_gint/grid_meshk.h:
```diff
@@ -19,9 +19,12 @@ class Grid_MeshK
 
-	protected:
+        /// move operator for the next ESolver to directly use its infomation
+        Grid_MeshK& operator=(Grid_MeshK&& rhs) = default;
```
I would suggest removing this line. We don't want operator unless necessary.

mohanchen commented on source/module_hamilt_lcao/module_gint/grid_technique.h:
```diff
@@ -22,6 +22,9 @@ class Grid_Technique : public Grid_MeshBall
     ~Grid_Technique();
+
+    /// move operator for the next ESolver to directly use its infomation
+    Grid_Technique& operator=(Grid_Technique&& rhs) = default;
```
I would suggest removing this line. We don't want operator unless necessary.

maki49 committed:
add integrate test

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes
